staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/list_width.go

```diff
@@ -0,0 +1,122 @@
+)
+
+func newListWidthEstimator(countFn objectCountGetterFunc) WidthEstimatorFunc {
+	estimator := &listWidthEstimator{
+		countFn: countFn,
+	}
+	return estimator.estimate
+}
+
+type listWidthEstimator struct {
+	countFn objectCountGetterFunc
+}
+
+func (e *listWidthEstimator) estimate(r *http.Request) Width {
+	requestInfo, ok := apirequest.RequestInfoFrom(r.Context())
+	if !ok {
+		// no RequestInfo should never happen, but to be on the safe side
+		// let's return minimumSeatsList
+		return Width{Seats: maximumSeats}
+	}
+
+	query := r.URL.Query()
+	listOptions := metav1.ListOptions{}
+	if err := metav1.Convert_url_Values_To_v1_ListOptions(&query, &listOptions, nil); err != nil {
+		klog.ErrorS(err, "Failed to convert options while calculating request width")
+		return Width{Seats: minimumSeats}
+	}
+
+	count := e.countFn.Get(key(requestInfo))
+	isListFromCache := !shouldListFromStorage(query, &listOptions)
+
+	// TODO: For resources that implement indexes at the watchcache level,
+	//  we need to adjust the cost accordingly
+	var estimatedObjectsToBeProcessed int64
+	switch {
+	case isListFromCache:
+		estimatedObjectsToBeProcessed = count
+	default:
+		// Even if a selector is specified and we may need to list and go over more objects from etcd
+		// to produce the result of size <limit>, each individual chunk will be of size at most <limit>.
+		// As a result. the width of the request should be computed based on <limit> and the actual
+		// cost of processing more elements will be hidden in the request processing latency.
+		estimatedObjectsToBeProcessed = listOptions.Limit
+	}
+
+	klog.Infof("estimatedObjectsToBeProcessed=%d count=%d, key=%s isListFromCache=%t", estimatedObjectsToBeProcessed, count, key(requestInfo), isListFromCache)
+
+	// for now, our rough estimate is to allocate one seat to each 100 obejcts that
+	// will be processed by the list request.
+	// we will come up with a different formula for the transformation function and/or
+	// fine tune this number in future iteratons.
+	seats := uint(math.Ceil(float64(estimatedObjectsToBeProcessed) / float64(100)))
+
+	// make sure we never return a seat of zero
+	if seats < minimumSeats {
+		seats = minimumSeats
+	}
+	if seats > maximumSeats {
+		seats = maximumSeats
+	}
+	return Width{Seats: seats}
+}
+
+func key(requestInfo *apirequest.RequestInfo) string {
+	groupResource := &schema.GroupResource{
+		Group:    requestInfo.APIGroup,
+		Resource: requestInfo.Resource,
+	}
+	return groupResource.String()
+}
+
+func shouldListFromStorage(query url.Values, opts *metav1.ListOptions) bool {
+	resourceVersion := opts.ResourceVersion
+	pagingEnabled := utilfeature.DefaultFeatureGate.Enabled(features.APIListChunking)
+	hasContinuation := pagingEnabled && len(opts.Continue) > 0
+	hasLimit := pagingEnabled && opts.Limit > 0 && resourceVersion != "0"
+	return resourceVersionNotSet(query) || hasContinuation || hasLimit || opts.ResourceVersionMatch == metav1.ResourceVersionMatchExact
```

Comment:
I had to slightly change it from `resourceVersion == ""` to `esourceVersionNotSet(query)` since `ListOptions.ResourceVersion` is empty by default. I tried to explain below. Please check if my assessment is correct.

Reply:
Sorry - I don't follow.

ResourceVersion should be empty by default - you need to explicitly set it (as a client) to have that set.

TBH - I don't understand why you need this function.

Reply:
[In other word - list from etcd (empty RV is a default)].